### ISSUE TITLE
Very different recommendations across machines

### ISSUE THREAD

**Issue by drraghuram** [lyst/lightfm#225]
With the same set of parameters, i.e., random_state, learning_rate, user_alpha, loss function, item_alpha, no_components, item_alpha and epochs, the recommendations that we are getting differ from each other on different machines.

` model = LightFM(loss='warp',
                    random_state=2016,
                    no_components= 20,
                    learning_rate=0.0350437109875,
                    user_alpha=7.88850830743e-05,
                    item_alpha=4.31495677869e-05)
    model.fit_partial(
            Mui_tr,
            epochs=10,
            verbose=True)
    return model`

The relevant code block is shown above. 
 The documentation on numpy.random.RandomState says the following, "_Compatibility Guarantee_: **A fixed seed and a fixed series of calls to ‘RandomState’ methods using the same parameters will always produce the same results up to roundoff error except when the values were incorrect.**"

**Comment by maciejkula** [comment]
1. Is the input data the same?
2. Are you using multithreaded fitting? If so, results are not deterministic.
3. Can you try comparing runs on the same machine?

Do the recommendations differ in a material way?

**Comment by drraghuram** [comment]
1. Yes, the input data is the same.
2. No, we are not using multithreaded fitting.
3. Repeated runs on the same machine gives the same reults. 

I am not sure what you mean by a material way, but the order in which the results appear does change. For example, a recommendation appearing as the 4th recommendation in one machine appears as the 6th recommendation in a different machine.

**Comment by maciejkula** [comment]
Is it possible that you use different versions of Python or numpy on the
different machines, or different C compilers?

**Comment by drraghuram** [comment]
All of us are using Python 2.7, with the same versions of numpy. The C compiler is the same across machines.

**Comment by maciejkula** [comment]
Interesting. Could you try initializing a `np.random.RandomState` with the same seed on the different machines and getting a sequence of random values? They should all be the same. If that holds, then we know the bug is in LightFM.

**Comment by drraghuram** [comment]
Hey, sorry for the late reply. Was unwell. We have already tried first. We got the same sequence of random values. Will try it again today to double check.

**Comment by maciejkula** [comment]
Hmm, there are two other possibilities.
1. Are you passing COO matrices to `fit`? If they are not COO, they are converted to COO, and that conversion may result in different order of elements, which affects the result. Similarly, if you are passing COO matrices, can you confirm that the order of entries is the same?
2. Are the CPU architectures of your machines the same? By default, LightFM is compiled with `--ffast-math`, which means that float operations are reordered and vectorized. You could test this by recompiling the dynamic objects without `--ffast-math`.

Out of curiosity, why is it important that the result be the same across machines? Usually, fitting is done once on a single machines, and only prediction is distributed.

**Comment by arita37** [comment]
In that topic,
1) is CSR input fine ? (or only COO)

2) how LFM recognize the order of userid ibetween interaction matrix and user_id ?

**Comment by maciejkula** [comment]
1. CSR input is fine, it will be transparently converted (but you'll use less memory if you start with the right format).
2. The row number of the interaction matrix is the user id.

**Comment by arita37** [comment]
When creating user feature and item feature,
we need to be careful on the “implicit Id”.


I have a training set, when using user feature and item feature (price,...), I have
both lower results on test and train sets.


Thanks.




On Dec 13, 2017, at 23:21, Maciej Kula <notifications@github.com> wrote:

CSR input is fine, it will be transparently converted (but you'll use less memory if you start with the right format).
The row number of the interaction matrix is the user id.
―

**Comment by maciejkula** [comment]
It's not really implicit: the i-th row of the interaction matrix (the i-th user) will use the features stored in the i-th row of the feature matrix. Have a look at the [Notes](http://lyst.github.io/lightfm/docs/lightfm.html#lightfm.LightFM) section in the documentation.

**Comment by maciejkula** [comment]
If your hybrid model performs worse than a pure CF model, make sure:
1. Your features are [appropriately scaled](https://en.wikipedia.org/wiki/Feature_scaling).
2. If your features are continuous, consider [discretization](https://en.wikipedia.org/wiki/Discretization_of_continuous_features).
3. You don't drop per-item or per-user features. Again, have a look at the [notes section](http://lyst.github.io/lightfm/docs/lightfm.html#lightfm.LightFM).

**Comment by arita37** [comment]
Am investigatin the input mapping :

Input interaction matrix:
userid itemid
7          14
5           11


Input item matrix:
11   ....
14. ....
Is this one correct/incorrect ?
(it seems we need to follow the interaction matrix order  for user/item ....)


Thanks !










On Dec 13, 2017, at 23:32, Maciej Kula <notifications@github.com> wrote:

If your hybrid model performs worse than a pure CF model, make sure:

Your features are appropriately scaled.
If your features are continuous, consider discretization.
You don't drop per-item or per-user features. Again, have a look at the notes section.
―

**Comment by maciejkula** [comment]
Item id 14 will look up the 14th row of the item feature matrix and so on.

**Comment by arita37** [comment]
Ok, am checking it.

On another item, after reading your paper,

1) why a full FM model is less convenient
that the dot product of item-user ?
(the model used in light).

2) Is there a way to add contextual feature,
ie attached to the pair (user, item) and not necessary to one of them ?









On Dec 14, 2017, at 8:26, Maciej Kula <notifications@github.com> wrote:

Item id 14 will look up the 14th row of the item feature matrix and so on.

―
